eclipse-jdt/eclipse.jdt.ui
# Implemented Batch SelfEncapsulateField for multiple fields at once #1080

Issue: https://github.com/eclipse-jdt/eclipse.jdt.ui/issues/1080



## What it does

Currently with Refactoring -> Encapsulate, encapsulating only one field at a time is possible. With this PR, it allows to select multiple fields and perform encapsulation on them in a batch.

## How to test

Select multiple fields at a time and click on Refactor -> Encapsulate fields in the menu. A wizard would appear to select the fields and specify how to perform the operation. Once done, click okay. This should encapsulate all the selected fields.

![image](https://github.com/eclipse-jdt/eclipse.jdt.ui/assets/46150646/b3e418ab-61e4-4959-8ff8-aaa22d4cc638)

## How it works
I tried many ways to implement this. The first approach was to select all the fields needed to be refactored and then roll out the operation in a loop. However, multiple change operations were registered in the undo manager and thus it required multiple undoing to undo the result. Another approach was to create a composite refactoring which consists the information of all the single field refactoring and create a composite change consisting of all the changes from each refactoring. However, the changes were still rolled out one by one since their textChange object were different for each refactoring. And thus changes made by one refactoring would change something in the file and the next refactoring's change object will lose the context of the field which has already been registered in its TextChange object. Thus, it would make faulty changes.

I tried a few more approaches until I realized, that all the refactoring use their ASTRewrite per compilationUnit to register their changes and then create a TextEdit from all the changes registered in the ASTRewrite. My idea was to share the ASTRewrite, ImportRewrite, ASTs and several resources which are needed to create changes among all the SelfEncapsulateFieldRefactorings. These resources are managed by the composite refactoring and the SelfEncapsulateFieldRefactorings can use the relevant resources and write their changes there. Once every refactoring has finished their changes, we use these resources in the composite refactoring to create a single Change Object per compilation unit to roll out the refactoring changes. Thus, it also registers only one Undo Operation.

## Limitations
**Update: This limitation no longer exists. At first we collect all the getter setter changes and then we change the access modifiers for all the fields.**
Since we pass the insertion index of the getter and setter to the rewriter, in a case where we have multiple fields in a declaration like int i, j; together, they are split in 2 lines like private int i; private int j;. It may change the insertion index of the getter and setter. However, this behaviour doesn't end to any failure but only the getter and setter are inserted at a wrong index (one index before the one where they are supposed to be inserted).


## Author checklist
